pkg/topology/full/full_test.go

```diff
@@ -0,0 +1,185 @@
+		err := fullDriver.AddPeer(context.Background(), overlay)
+		if !errors.Is(err, topology.ErrNotFound) {
+			t.Fatalf("full conn driver returned err %s", err.Error())
+		}
+
+		if discovery.Broadcasts() != 0 {
+			t.Fatalf("broadcasts expected %v, got %v ", 0, discovery.Broadcasts())
+		}
+	})
+
+	t.Run("OK - connected peers - peer already connected", func(t *testing.T) {
+		discovery := mock.NewDiscovery()
+		addressbook := inmem.New()
+		alreadyConnected := connectedPeers[0].Address
+
+		p2p := p2pmock.New(p2pmock.WithConnectFunc(func(ctx context.Context, addr ma.Multiaddr) (swarm.Address, error) {
+			t.Fatal("should not be called")
+			return swarm.Address{}, nil
+		}), p2pmock.WithPeersFunc(func() []p2p.Peer {
+			return connectedPeers
+		}))
+
+		fullDriver := full.New(discovery, addressbook, p2p, logger)
+		multiaddr, err := ma.NewMultiaddr(underlay)
+		if err != nil {
+			t.Fatal("error creating multiaddr")
+		}
+
+		addressbook.Put(alreadyConnected, multiaddr)
+		err = fullDriver.AddPeer(context.Background(), alreadyConnected)
+		if err != nil {
+			t.Fatalf("full conn driver returned err %s", err.Error())
+		}
+
+		if discovery.Broadcasts() != 3 {
+			t.Fatalf("broadcasts expected %v, got %v ", 3, discovery.Broadcasts())
+		}
+
+		// check newly added node
+		if err := checkAddreseeRecords(discovery, alreadyConnected, connectedPeers[1:]); err != nil {
+			t.Fatal(err)
+		}
+
+		// check other nodes
+		for _, p := range connectedPeers[1:] {
+			if err := checkAddreseeRecords(discovery, p.Address, connectedPeers[0:1]); err != nil {
+				t.Fatal(err)
+			}
+		}
+	})
+
+	t.Run("OK - connected peers - peer not already connected", func(t *testing.T) {
+		discovery := mock.NewDiscovery()
+		addressbook := inmem.New()
+
+		p2ps := p2pmock.New(p2pmock.WithConnectFunc(func(ctx context.Context, addr ma.Multiaddr) (swarm.Address, error) {
+			if addr.String() != underlay {
+				t.Fatalf("expected multiaddr %s, got %s", addr.String(), underlay)
+			}
+			return overlay, nil
+		}), p2pmock.WithPeersFunc(func() []p2p.Peer {
+			return connectedPeers
+		}))
+
+		fullDriver := full.New(discovery, addressbook, p2ps, logger)
+		multiaddr, err := ma.NewMultiaddr(underlay)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		addressbook.Put(overlay, multiaddr)
+		err = fullDriver.AddPeer(context.Background(), overlay)
+		if err != nil {
+			t.Fatalf("full conn driver returned err %s", err.Error())
+		}
+
+		if discovery.Broadcasts() != 4 {
+			t.Fatalf("broadcasts expected %v, got %v ", 4, discovery.Broadcasts())
+		}
+
+		// check newly added node
+		if err := checkAddreseeRecords(discovery, overlay, connectedPeers); err != nil {
+			t.Fatal(err)
+		}
+
+		// check other nodes
+		for _, p := range connectedPeers {
+			if err := checkAddreseeRecords(discovery, p.Address, []p2p.Peer{{Address: overlay}}); err != nil {
+				t.Fatal(err)
+			}
+		}
+	})
+}
+
+func checkAddreseeRecords(discovery *mock.Discovery, addr swarm.Address, expected []p2p.Peer) error {
```

Comment:
you can change the signature to also accept `t *testing.T`, then declare `t.Helper()` at the head of the function, and then you can directly `t.Fatal` or `t.Errorf` inside the helper

Reply:
I feel like this makes more room to use this functions in different places. Do you think it has any downsides?